core/mempool/src/lib.rs

```diff
@@ -300,46 +395,95 @@ impl Pool {
             if remove_map {
                 guard.remove(&sender);
             }
+            info_guard.remove(&transaction.get_hash());
         }
         for receipt in block.body.receipts.iter() {
             self.receipts.write().expect(POISONED_LOCK_ERR).remove(receipt);
         }
     }
 
-    pub fn add_pending(&self, authority_id: AuthorityId, hash: CryptoHash) {
-        self.pending_snapshots.write().expect(POISONED_LOCK_ERR).push((authority_id, hash))
+    pub fn fetch_payload(&self, request: MissingPayloadRequest) -> MissingPayloadResponse {
+        let transactions: Vec<_> = {
+            let guard = self.transactions.read().expect(POISONED_LOCK_ERR);
+            let info_guard = self.transaction_info.read().expect(POISONED_LOCK_ERR);
+            request
+                .transactions
+                .into_iter()
+                .filter_map(|hash| {
+                    info_guard
+                        .get(&hash)
+                        .and_then(|(id, nonce)| guard.get(id).and_then(|map| map.get(nonce)))
+                        .cloned()
+                })
+                .collect()
+        };
+        let receipts: Vec<_> = {
+            let guard = self.receipts.read().expect(POISONED_LOCK_ERR);
+            request.receipts.into_iter().filter_map(|hash| guard.get(&hash).cloned()).collect()
+        };
+
+        MissingPayloadResponse { transactions, receipts, snapshot_hash: request.snapshot_hash }
     }
 
+    /// add snapshot sent by peer to pending_snapshots and request the missing
```

Comment:
The docstring says that it adds snapshot to `pending_snapshots`, but it only adds them to `pending_snaphots` when some transactions/receipts are missing, otherwise it adds it to `snapshots`. Please fix the docstring.